File: src/Day2RockPaperScissors/c/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <direct.h>

void concatenate_string(char *s, char *s1)
{
    int i;
    int j = strlen(s);

    for (i = 0; s1[i] != '\0'; i++)
    {
        s[i + j] = s1[i];
    }
    s[i + j] = '\0';
    return;
}

enum Shape
{
    Rock = 1,
    Paper = 2,
    Scissors = 3
};
const char *shape_to_string(enum Shape shape)
{
    switch (shape)
    {
    case Rock:
        return "Rock";
    case Paper:
        return "Paper";
    case Scissors:
        return "Scissors";
    }
}

enum Shape shape_from_input_value(char character)
{
    switch (character)
    {
    case 'A':
    case 'X':
        return Rock;
    case 'B':
    case 'Y':
        return Paper;
    case 'C':
    case 'Z':
        return Scissors;
    }
}

int score_from_shape(enum Shape shape)
{
    switch (shape)
    {
    case Rock:
        return 1;
    case Paper:
        return 2;
    case Scissors:
        return 3;
    }
}

// loss = 0
// draw = 3
// win = 6
int score_from_result(enum Shape p1Shape, enum Shape p2Shape)
{
    if (p1Shape == p2Shape)
    {
        return 3;
    }

    if ((p1Shape == Rock && p2Shape == Scissors) ||
        (p1Shape == Scissors && p2Shape == Paper) ||
        (p1Shape == Paper && p2Shape == Rock))
    {
        return 6;
    }

    return 0;
}

int shape_from_expected_score(char result, enum Shape opponent)
{
    if (result == 'Y')
        return opponent;

    if (opponent == Scissors)
    {
        switch (result)
        {
        case 'X':
            return Paper; // Loss
        case 'Z':
            return Rock; // Victory
        }
    }

    if (opponent == Rock)
    {
        switch (result)
        {
        case 'X':
            return Scissors; // Loss
        case 'Z':
            return Paper; // Victory
        }
    }

    if (opponent == Paper)
    {
        switch (result)
        {
        case 'X':
            return Rock; // Loss
        case 'Z':
            return Scissors; // Victory
        }
    }
}

int main()
{
    // Get input file path
    char filePath[256];
    _getcwd(filePath, 256);
    concatenate_string(filePath, "\\input.txt");
    // printf("File path: \n%s\n", filePath);

    // Task 1
    printf("Task 1:\n");
    FILE *fpointer = fopen(filePath, "r"); // r file mode to read contents of the files
    char line[20];

    int totalScore = 0;
    while (fgets(line, sizeof(line), fpointer) != NULL)
    {
        if (strlen(line) > 1) // if this line has chars
        {
            enum Shape opponentShape = shape_from_input_value(line[0]);
            enum Shape playerShape = shape_from_input_value(line[2]);

            totalScore += score_from_shape(playerShape);
            totalScore += score_from_result(playerShape, opponentShape);
        }
    }
    printf("Answer: %i\n", totalScore);

    fclose(fpointer);

    // Task 2
    printf("Task 2:\n");
    FILE *fpointer2 = fopen(filePath, "r"); // r file mode to read contents of the files
    char line2[20];

    while (fgets(line2, sizeof(line2), fpointer2) != NULL)
    {
        if (strlen(line2) > 1) // if this line has chars
        {
            enum Shape opponentShape = shape_from_input_value(line2[0]);
            enum Shape playerShape = shape_from_expected_score(line2[2], opponentShape);

            totalScore += score_from_shape(playerShape);
            totalScore += score_from_result(playerShape, opponentShape);
        }
    }
    printf("Answer: %i\n", totalScore);

    fclose(fpointer2);

    return 0;
}

File: src/Day2RockPaperScissors/rs/src/main.rs
use std::env;
use std::fs;

fn main() {
    // env::set_var("RUST_BACKTRACE", "1");
    task1();
    task2();
}

fn task1() {
    println!("Task 1:");

    // Get input file
    let current_dir = env::current_dir().unwrap();
    let parent_dir = current_dir.parent().unwrap();
    let file_path = format!("{}{}", parent_dir.display(), "\\input.txt");

    let input_file = fs::read_to_string(file_path).unwrap();
    let lines: Vec<&str> = input_file.split('\n').collect();

    let mut result: i16 = 0;

    for line in lines {
        let o_char = line.chars().nth(0).unwrap();
        let p_char = line.chars().nth(2).unwrap();
        let opponent_shape = char_to_shape(o_char);
        let player_shape = char_to_shape(p_char);

        result += round_to_points(player_shape, opponent_shape);
    }

    println!("Result: {}", result);
}

fn task2() {
    println!("Task 2:");

    // Get input file
    let current_dir = env::current_dir().unwrap();
    let parent_dir = current_dir.parent().unwrap();
    let file_path = format!("{}{}", parent_dir.display(), "\\input.txt");

    let input_file = fs::read_to_string(file_path).unwrap();
    let lines: Vec<&str> = input_file.split('\n').collect();

    let mut result: i16 = 0;

    for line in lines {
        let p_char = line.chars().nth(2).unwrap();
        let o_char = line.chars().nth(0).unwrap();
        let opponent_shape = char_to_shape(o_char);
        let (player_shape, opponent_shape) = shape_from_expected_score(p_char, opponent_shape);

        result += round_to_points(player_shape, opponent_shape);
    }

    println!("Result: {}", result);
}

#[derive(PartialEq, Clone)]
enum Shape {
    Rock = 1,
    Paper = 2,
    Scissors = 3,
}
impl std::fmt::Display for Shape {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        match self {
            Shape::Rock => write!(f, "Rock"),
            Shape::Paper => write!(f, "Paper"),
            Shape::Scissors => write!(f, "Scissors"),
        }
    }
}

fn char_to_shape(character: char) -> Shape {
    return match character {
        'A' | 'X' => Shape::Rock,
        'B' | 'Y' => Shape::Paper,
        'C' | 'Z' => Shape::Scissors,
        _ => panic!("Unidentified char passed into method: {}", character),
    };
}

fn round_to_points(player_shape: Shape, opponent_shape: Shape) -> i16 {
    let mut points: i16 = 0;

    // number of points from the player shape
    match player_shape {
        Shape::Rock => points += 1,
        Shape::Paper => points += 2,
        Shape::Scissors => points += 3,
    };

    // number of points from the result
    if player_shape == opponent_shape {
        points += 3;
    }

    if (player_shape == Shape::Rock && opponent_shape == Shape::Scissors)
        || (player_shape == Shape::Scissors && opponent_shape == Shape::Paper)
        || (player_shape == Shape::Paper && opponent_shape == Shape::Rock)
    {
        points += 6;
    }

    return points;
}

fn shape_from_expected_score(expected_score: char, opponent_shape: Shape) -> (Shape, Shape) {
    let player_shape: Shape;

    if expected_score == 'Y' {
        player_shape = opponent_shape.clone();
        return (player_shape, opponent_shape);
    }

    if opponent_shape == Shape::Scissors {
        match expected_score {
            'X' => player_shape = Shape::Paper, // Loss
            'Z' => player_shape = Shape::Rock,  // Victory
            _ => panic!(
                "Unidentified expected score passed into method: {}",
                expected_score
            ),
        }
    } else if opponent_shape == Shape::Rock {
        match expected_score {
            'X' => player_shape = Shape::Scissors, // Loss
            'Z' => player_shape = Shape::Paper,    // Victory
            _ => panic!(
                "Unidentified expected score passed into method: {}",
                expected_score
            ),
        }
    } else if opponent_shape == Shape::Paper {
        match expected_score {
            'X' => player_shape = Shape::Rock,     // Loss
            'Z' => player_shape = Shape::Scissors, // Victory
            _ => panic!(
                "Unidentified expected score passed into method: {}",
                expected_score
            ),
        }
    } else {
        panic!("Unrecognised shape passed into method");
    }

    return (player_shape, opponent_shape);
}
